### PR TITLE
[ReadyforReview]Fix empty input in TrainingDatum and add test

### LINKED ISSUE #339: Fix empty `Input`s being allowed in `TrainingDatum`
### Motivation
------------------
There is a small bug on the edge case of `Input` construction which allows the creation of an empty `Input` class. Given that `Input` is its own type it seems fair that having an empty `Input` is reasonable. (You can have empty lists, tuples, MultiLevels etc.). However, within a `TrainingDatum` that must have a `Real` output attached to it, an empty `Input` makes no mathematical sense.  

For example:
```python

# Raises TypeError
test_input1 = TrainingDatum(Input(None), 1)

# Raises no error and can pass through
test_input2 = TrainingDatum(Input(), 1)
```

### Acceptance Criteria 
----------------------------
I would suggest making an alteration to the `_validate_input` static method in the `TrainingDatum` class, to also throw an error for the length of the input <1.

An example would be: 

```python
    @staticmethod
    def _validate_input(input: Any) -> None:
        """Check that an object is an instance of an Input with dimensions >= 1, 
             raising the appropriate error if not."""

        if not isinstance(input, Input):
            raise TypeError("Argument 'input' must be of type Input")
      
        if len(input) < 1: 
             raise ValueError("Input requires dimensions >= 1")

```
This should have an appropriate test as well as well as an alteration to the docstring. 

Note: 
This may be a design choice to discuss with TH, as it may be that simply not allowing the creation of Input() is more suitable.

### PR DESCRIPTION
**Pull request by HarryWhiteRSE**
Closes #339 

`TrainingDatum` will now raise a `ValueError` upon construction if the `Input` passed to it is empty. `Input` itself does error checking for types upon construction and `TrainingDatum` ensures that the type being passed is an `Input`. This seemed like the most efficient and useful fix as still allowing an empty `Input` type maybe useful if there is ever the option of features similar to lists like `append` etc. 

_Note to reviewer: This is a good first basic PR to review to show how an issue was created from an edge case bug, albeit small, and a fix can be implemented using that edge case from the issue as the basis for the test case._

### CODE REVIEW

**Review by mbjohns — APPROVED** [review]
Good work! 🚀

File: tests/unit/test_modelling.py [hunk 1]
```diff
@@ -305,6 +305,14 @@ def test_input_error(self):
         with self.assertRaisesRegex(TypeError, exact(msg)):
             TrainingDatum(1, 1)
 
+    def test_empty_input_error(self):
+        """Test that a ValueError is raised if the constructor arg 'input'
+        is an empty Input."""
+
+        msg = "Argument 'input' is empty, it must contain a value."
```
Comment (mbjohns) [inline tests/unit/test_modelling.py:312]: We changed the message but not in the test 🤦

**Review by mbjohns — CHANGES_REQUESTED** [review]

**Review by mbjohns — APPROVED** [review]
All tests pass my end :) 🚀

File: tests/unit/test_modelling.py (hunk 1 above)
Comment (HarryWhiteRSE) [reply tests/unit/test_modelling.py]: Oopsies 🤦🏼
